### PR TITLE
MRG: Epochs API

### PR DESCRIPTION
**Pull request by agramfort**
the idea is to remove the work epochs in a method that belongs to the Epochs class.

I find it cleaner this way but you can object :)

### CODE REVIEW

File: doc/manual/memory.rst [hunk 1]
```diff
@@ -38,7 +38,7 @@ has been loaded with ``preload=True``. Preloading is also supported for :func:`m
 
 .. warning:: This comes with a caveat. When ``preload=False``, data rejection based on peak-to-peak thresholds is executed when the data is loaded from disk, *not* when the ``Epochs`` object is created.
 
-To explicitly reject artifacts with ``preload=False``, use the function :func:`mne.Epochs.drop_bad_epochs`.
+To explicitly reject artifacts with ``preload=False``, use the function :func:`mne.Epochs.drop_bad`.
```
Comment (dengemann) [inline doc/manual/memory.rst]: should be *bads by convention, no?

**Comment by dengemann** [comment]
Other than that I like it!

**Comment by agramfort** [comment]
drop_bads makes sense to me.

any other opinion?

**Comment by Eric89GXL** [comment]
`drop_bad` is fine, but how about `drop_indices` instead of just `drop`? It's weird to have one `drop_*` and one that is just `drop`.

**Comment by agramfort** [comment]
can drop by mistaken by something else?

if we drop the drop naming ( :) ) then I prefer drop_by_index

I'll let the others voice their opinion...

**Comment by jona-sassenhagen** [comment]
"bads" elsewhere refers to channels, doesn't it? E.g. info["bads"]. So is it universally intuitive that drop_bad(s) drops epochs and not channels?

**Comment by agramfort** [comment]
or shall we fuse the two methods into a single drop function?

indices could be None which would mean just drop based on reject?

**Comment by Eric89GXL** [comment]
Maybe `drop_reject` and `drop_by_index`? A single function I think is a little bit less clear

**Comment by Eric89GXL** [comment]
> can drop by mistaken by something else?

Well there are currently two types of `drop`-like functions being discussed. I don't see why one should be known as just `drop` and the other `drop_by_index`. Or to put it conversely, why one should be called `drop` and the other way `drop_reject`.

**Comment by teonbrooks** [comment]
I like having one `drop` method IMO

**Comment by agramfort** [comment]
> I like having one drop method IMO
> 

Any thought on API?

**Comment by teonbrooks** [comment]
I don't see where `drop_bad_epochs` is updating the `drop_log`. If this were to work as a unified api, I would suggest `drop(indices=[], reject='existing', flat='existing', reason=None , verbose=None) and internally if reason is None and indices is a non-empty list, the reason defaults to 'USER'.

**Comment by agramfort** [comment]
>
> I don't see where drop_bad_epochs is updating the drop_log. If this were
> to work as a unified api, I would suggest `drop(indices=[],
> reject='existing', flat='existing', reason=None , verbose=None) and
> internally if reason is None and indices is a non-empty list, the reason
> defaults to 'USER'.
>

sounds clean to me. Any other view on this?

**Comment by jaeilepp** [comment]
My vote goes for ``drop_bads`` and ``drop_by_index``. These two seem the most clean and intuitive to me. I don't like the idea of unifying the two methods.

**Comment by Eric89GXL** [comment]
I can live with unification as long as it's clearly laid out the order in which the operations are applied, e.g. `drop([1, 3])` will first drop the first and third epochs/events, and then do reject (and flat and annotation) based dropping.

**Comment by wmvanvliet** [comment]
Why would we break API over this?
I vote for:

`drop_bad_epochs` (as it was)
`drop` (new function that drop the epochs with the supplied indices)

**Comment by Eric89GXL** [comment]
We should settle on this before 0.12, @agramfort do you see a consensus?

**Comment by Eric89GXL** [comment]
Also you need a rebase, if you are too busy I can take over once you sort out the naming

**Comment by agramfort** [comment]
well I think that

epochs.drop_bad() is clear that we drop bad epochs.

and then

epochs.drop

is like the pandas.DataFrame.drop() method...

I kind of like what I proposed in the end...

**Comment by agramfort** [comment]
note that the use of drop methods should be documented in tutorials/plot_object_epochs.py :)

**Comment by Eric89GXL** [comment]
What about the unified `drop(indices=[], reject='existing', flat='existing', reason=None , verbose=None)`? It fits with `pandas`, unifies behavior, and cleans up the names all at once.

**Comment by agramfort** [comment]
well then when you set indices reject and flat are ignored params?

**Comment by Eric89GXL** [comment]
The order would be 1) drop indices 2) drop based on reject/flat. So `epochs.drop(idx)` would drop the indices and then do reject/flat based rejection. Alternatively we could do:
```
drop(indices=[], reject='auto', flat='auto', reason=None, verbose=None)
```
Which would mean `epochs.drop()` would drop based on reject/flat only, and `epochs.drop(idx)` would drop based on reject/flat only, and `epochs.drop(idx, 'existing', 'existing')` would drop based on indices, then reject and flat.

**Comment by dengemann** [comment]
drop_bad and drop_indices would be fundamentally different. The one
executes previously made decisions about bad epochs. The other would be
more pandas like in the sense of taking some some of index. I slightly
favor not mixing these patterns. To me it's easier to memorize 2 simple but
complementary expressions than 1 complex expression.

**Comment by agramfort** [comment]
same feeling as @dengemann

I vote for 2 methods. I find it easier to memorize and explain.

I would go for drop_bad and drop.

**Comment by Eric89GXL** [comment]
Actually I don't like my `auto` suggestion now that I think about it, automated processes would then have to triage based on whether or not they had `len(idx)==0` which is ugly...

Sounds like `drop` and `drop_bad` is starting to win from a Python-wide-unification perspective (didn't realize `pandas` had that method originally), which I can live with. @teonbrooks @wmvanvliet @jaeilepp @jona-sassenhagen okay with you, too?

**Comment by jaeilepp** [comment]
Fine with me, as long as the docs clearly state that we're dropping epochs and not channels or something.

**Comment by agramfort** [comment]
Eric I'll let you take over.

and don't forget to document it in tutorials/plot_object_epochs.py :)

**Comment by Eric89GXL** [comment]
[Tutorial updated] (https://circle-artifacts.com/gh/mne-tools/mne-python/751/artifacts/0/home/ubuntu/mne-python/doc/_build/html/auto_tutorials/plot_object_epochs.html) and [docstrings here] (https://circle-artifacts.com/gh/mne-tools/mne-python/751/artifacts/0/home/ubuntu/mne-python/doc/_build/html/generated/mne.Epochs.html#mne.Epochs.drop).

Ready for review/merge from my end.

File: mne/epochs.py [hunk 2]
```diff
@@ -1233,16 +1250,24 @@ def plot_image(self, picks=None, sigma=0., vmin=None,
                                  cmap=cmap, fig=fig,
                                  overlay_times=overlay_times)
 
-    @verbose
+    @deprecated('drop_bad method has been renamed drop_bad. '
+                'drop_bad method will be removed in 0.13')
```
Comment (jaeilepp) [inline mne/epochs.py]: -> ``'drop_epochs method has been renamed drop_bad. '
 'drop_epochs method will be removed in 0.13'``

**Comment by jaeilepp** [comment]
Otherwise LGTM

File: tutorials/plot_object_epochs.py [hunk 3]
```diff
@@ -87,6 +87,19 @@
...
+# thresholds with :func:`epochs.drop_bad(reject, flat) <mne.Epochs.drop_bad>`.
+# You can also inspect the reason why epochs were dropped by looking at the
+# list stored in ``epochs.drop_log`` or plot them with
+# :func:`epochs.plot_drop_log() <mne.Epochs.plot_drop_log>`.
+
+epochs.drop([0], reason='User reason')
+epochs.drop_bad(reject=dict(grad=2500e-13, mag=4e-12, eog=200e-6), flat=None)
+print(epochs.drop_log)
+epochs.plot_drop_log()
+
```
Comment (agramfort) [inline tutorials/plot_object_epochs.py]: Can you document also the selection attribute?

**Comment by Eric89GXL** [comment]
Comments addressed

**Comment by jaeilepp** [comment]
LGTM

**Comment by agramfort** [comment]
Merge when CI is happy

File: doc/manual/memory.rst (hunk 1 above)
Comment (teonbrooks) [inline doc/manual/memory.rst]: did we address this comment of `*_bad` vs. `*_bads`?

Comment (Eric89GXL) [inline doc/manual/memory.rst]: @dengemann was involved in the main thread where the names were brought up again as `drop_bad`, so I think so

**Pull request merged**